File: dmn-test-cases/standard/tck/1.3/cl3/0085-decision-services/translator/expected/java/dmn/DecisionService_003.java

import java.util.*;
import java.util.stream.Collectors;

@javax.annotation.Generated(value = {"ds.ftl", "decisionService_003"})
@com.gs.dmn.runtime.annotation.DRGElement(
    namespace = "",
    name = "decisionService_003",
    label = "",
    elementKind = com.gs.dmn.runtime.annotation.DRGElementKind.DECISION_SERVICE,
    expressionKind = com.gs.dmn.runtime.annotation.ExpressionKind.OTHER,
    hitPolicy = com.gs.dmn.runtime.annotation.HitPolicy.UNKNOWN,
    rulesCount = -1
)
public class DecisionService_003 extends com.gs.dmn.runtime.JavaTimeDMNBaseDecision<String> {
    public static final com.gs.dmn.runtime.listener.DRGElement DRG_ELEMENT_METADATA = new com.gs.dmn.runtime.listener.DRGElement(
        "",
        "decisionService_003",
        "",
        com.gs.dmn.runtime.annotation.DRGElementKind.DECISION_SERVICE,
        com.gs.dmn.runtime.annotation.ExpressionKind.OTHER,
        com.gs.dmn.runtime.annotation.HitPolicy.UNKNOWN,
        -1
    );

    private static class DecisionService_003LazyHolder {
        static final DecisionService_003 INSTANCE = new DecisionService_003();
    }
    public static DecisionService_003 instance() {
        return DecisionService_003LazyHolder.INSTANCE;
    }

    private final Decision_003 decision_003;

    private DecisionService_003() {
        this(new Decision_003());
    }

    private DecisionService_003(Decision_003 decision_003) {
        this.decision_003 = decision_003;
    }

    @java.lang.Override()
    public String applyMap(java.util.Map<String, String> input_, com.gs.dmn.runtime.ExecutionContext context_) {
        try {
            return apply(input_.get("inputData_003"), input_.get("decision_003_input_1"), input_.get("decision_003_input_2"), context_);
        } catch (Exception e) {
            logError("Cannot apply decision 'DecisionService_003'", e);
            return null;
        }
    }

    @java.lang.Override()
    public String applyPojo(com.gs.dmn.runtime.ExecutableDRGElementInput input_, com.gs.dmn.runtime.ExecutionContext context_) {
        try {
            return apply(((DecisionService_003Input_)input_).getInputData_003(), ((DecisionService_003Input_)input_).getDecision_003_input_1(), ((DecisionService_003Input_)input_).getDecision_003_input_2(), context_);
        } catch (Exception e) {
            logError("Cannot apply element 'DecisionService_003'", e);
            return null;
        }
    }

    @java.lang.Override()
    public String applyContext(com.gs.dmn.runtime.Context input_, com.gs.dmn.runtime.ExecutionContext context_) {
        try {
            return applyPojo(new DecisionService_003Input_(input_), context_);
        } catch (Exception e) {
            logError("Cannot apply element 'DecisionService_003'", e);
            return null;
        }
    }

    public String apply(String inputData_003, String decision_003_input_1, String decision_003_input_2, com.gs.dmn.runtime.ExecutionContext context_) {
        try {
            // Start DS 'decisionService_003'
            com.gs.dmn.runtime.annotation.AnnotationSet annotationSet_ = context_ != null ? context_.getAnnotations() : null;
            com.gs.dmn.runtime.listener.EventListener eventListener_ = context_ != null ? context_.getEventListener() : null;
            com.gs.dmn.runtime.external.ExternalFunctionExecutor externalExecutor_ = context_ != null ? context_.getExternalFunctionExecutor() : null;
            com.gs.dmn.runtime.cache.Cache cache_ = context_ != null ? context_.getCache() : null;
            long decisionService_003StartTime_ = System.currentTimeMillis();
            com.gs.dmn.runtime.listener.Arguments decisionService_003Arguments_ = new com.gs.dmn.runtime.listener.Arguments();
            decisionService_003Arguments_.put("inputData_003", inputData_003);
            decisionService_003Arguments_.put("decision_003_input_1", decision_003_input_1);
            decisionService_003Arguments_.put("decision_003_input_2", decision_003_input_2);
            eventListener_.startDRGElement(DRG_ELEMENT_METADATA, decisionService_003Arguments_);

            // Bind input decisions
            cache_.bind("decision_003_input_1", decision_003_input_1);
            cache_.bind("decision_003_input_2", decision_003_input_2);

            // Evaluate DS 'decisionService_003'
            String output_ = lambda.apply(inputData_003, decision_003_input_1, decision_003_input_2, context_);

            // End DS 'decisionService_003'
            eventListener_.endDRGElement(DRG_ELEMENT_METADATA, decisionService_003Arguments_, output_, (System.currentTimeMillis() - decisionService_003StartTime_));

            return output_;
        } catch (Exception e) {
            logError("Exception caught in 'decisionService_003' evaluation", e);
            return null;
        }
    }

    public com.gs.dmn.runtime.LambdaExpression<String> lambda =
        new com.gs.dmn.runtime.LambdaExpression<String>() {
            public String apply(Object... args_) {
                String inputData_003 = 0 < args_.length ? (String) args_[0] : null;
                String decision_003_input_1 = 1 < args_.length ? (String) args_[1] : null;
                String decision_003_input_2 = 2 < args_.length ? (String) args_[2] : null;
                com.gs.dmn.runtime.ExecutionContext context_ = 3 < args_.length ? (com.gs.dmn.runtime.ExecutionContext) args_[3] : null;
                com.gs.dmn.runtime.annotation.AnnotationSet annotationSet_ = context_ != null ? context_.getAnnotations() : null;
                com.gs.dmn.runtime.listener.EventListener eventListener_ = context_ != null ? context_.getEventListener() : null;
                com.gs.dmn.runtime.external.ExternalFunctionExecutor externalExecutor_ = context_ != null ? context_.getExternalFunctionExecutor() : null;
                com.gs.dmn.runtime.cache.Cache cache_ = context_ != null ? context_.getCache() : null;

                // Apply child decisions
                String decision_003 = DecisionService_003.this.decision_003.apply(inputData_003, context_);

                return decision_003;
            }
    };
}

File: dmn-test-cases/standard/tck/1.3/cl3/0085-decision-services/translator/expected/python/dmn/DecisionService_003.py
import typing
import decimal
import datetime
import isodate
import time

import jdmn.runtime.Context
import jdmn.runtime.DefaultDMNBaseDecision
import jdmn.runtime.ExecutionContext
import jdmn.runtime.LambdaExpression
import jdmn.runtime.LazyEval
import jdmn.runtime.Pair
import jdmn.runtime.Range
import jdmn.runtime.RuleOutput
import jdmn.runtime.RuleOutputList

import jdmn.runtime.annotation.Annotation
import jdmn.runtime.annotation.AnnotationSet
import jdmn.runtime.annotation.DRGElementKind
import jdmn.runtime.annotation.ExpressionKind
import jdmn.runtime.annotation.HitPolicy

import jdmn.runtime.cache.Cache

import jdmn.runtime.external.ExternalFunctionExecutor

import jdmn.runtime.listener.Arguments
import jdmn.runtime.listener.DRGElement
import jdmn.runtime.listener.EventListener
import jdmn.runtime.listener.Rule

import Decision_003


# Generated(value = {"ds.ftl", "decisionService_003"})
class DecisionService_003(jdmn.runtime.DefaultDMNBaseDecision.DefaultDMNBaseDecision):
    DRG_ELEMENT_METADATA: jdmn.runtime.listener.DRGElement.DRGElement = jdmn.runtime.listener.DRGElement.DRGElement(
        "",
        "decisionService_003",
        "",
        jdmn.runtime.annotation.DRGElementKind.DRGElementKind.DECISION_SERVICE,
        jdmn.runtime.annotation.ExpressionKind.ExpressionKind.OTHER,
        jdmn.runtime.annotation.HitPolicy.HitPolicy.UNKNOWN,
        -1
    )
    _instance = None

    def __init__(self):
        raise RuntimeError("Call instance() instead")

    @classmethod
    def instance(cls):
        if cls._instance is None:
            cls._instance = cls.__new__(cls)
            jdmn.runtime.DefaultDMNBaseDecision.DefaultDMNBaseDecision.__init__(cls._instance)
            cls._instance.initSubDecisions()
        return cls._instance

    def initSubDecisions(self, decision_003: Decision_003.Decision_003 = None):
        self.decision_003 = Decision_003.Decision_003() if decision_003 is None else decision_003

    def apply(self, inputData_003: typing.Optional[str], decision_003_input_1: typing.Optional[str], decision_003_input_2: typing.Optional[str], annotationSet_: jdmn.runtime.annotation.AnnotationSet.AnnotationSet, eventListener_: jdmn.runtime.listener.EventListener.EventListener, externalExecutor_: jdmn.runtime.external.ExternalFunctionExecutor.ExternalFunctionExecutor, cache_: jdmn.runtime.cache.Cache.Cache) -> typing.Optional[str]:
        try:
            # Start DS 'decisionService_003'
            decisionService_003StartTime_ = int(time.time_ns()/1000)
            decisionService_003Arguments_ = jdmn.runtime.listener.Arguments.Arguments()
            decisionService_003Arguments_.put("inputData_003", inputData_003)
            decisionService_003Arguments_.put("decision_003_input_1", decision_003_input_1)
            decisionService_003Arguments_.put("decision_003_input_2", decision_003_input_2)
            eventListener_.startDRGElement(self.DRG_ELEMENT_METADATA, decisionService_003Arguments_)

            # Bind input decisions
            cache_.bind("decision_003_input_1", decision_003_input_1)
            cache_.bind("decision_003_input_2", decision_003_input_2)

            # Evaluate DS 'decisionService_003'
            output_: typing.Optional[str] = self.evaluate(inputData_003, decision_003_input_1, decision_003_input_2, annotationSet_, eventListener_, externalExecutor_, cache_)

            # End DS 'decisionService_003'
            eventListener_.endDRGElement(self.DRG_ELEMENT_METADATA, decisionService_003Arguments_, output_, (int(time.time_ns()/1000) - decisionService_003StartTime_))

            return output_
        except Exception as e:
            self.logError("Exception caught in 'decisionService_003' evaluation", e)
            return None

    def evaluate(self, inputData_003: typing.Optional[str], decision_003_input_1: typing.Optional[str], decision_003_input_2: typing.Optional[str], annotationSet_: jdmn.runtime.annotation.AnnotationSet.AnnotationSet, eventListener_: jdmn.runtime.listener.EventListener.EventListener, externalExecutor_: jdmn.runtime.external.ExternalFunctionExecutor.ExternalFunctionExecutor, cache_: jdmn.runtime.cache.Cache.Cache) -> typing.Optional[str]:
        # Apply child decisions
        decision_003: typing.Optional[str] = self.decision_003.apply(inputData_003, annotationSet_, eventListener_, externalExecutor_, cache_)

        return decision_003
